### PR TITLE
chore: add federated learning example

### PR DESCRIPTION
**Pull request by fd0r**

### CODE REVIEW

**Review by bcm-at-zama — COMMENTED** [review]
Thx @fd0r . Would it be possible to have a single .md? I think it would be clearer. Also, there are client.py both in flower repo and here in the repo? it may be confusing, maybe take unique names to clarify

**Comment by fd0r** [comment]
> Thx @fd0r . Would it be possible to have a single .md? I think it would be clearer. Also, there are client.py both in flower repo and here in the repo? it may be confusing, maybe take unique names to clarify

Well I didn't want to remove the readme from the flower repository but I guess as long as we link it we are fine?
Or were you suggesting appending the current README to the end of the Flower README?

Yes I can indeed rename the client.py file so that it's more explicit.

File: use_case_examples/federated_learning/client.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        def evaluate(self, parameters, config):  # type: ignore
+            utils.set_model_params(model, parameters)
+            loss = log_loss(y_test, model.predict_proba(X_test))
+            accuracy = model.score(X_test, y_test)
+            return loss, len(X_test), {"accuracy": accuracy}
+
+    # Start Flower client
+    fl.client.start_numpy_client(server_address="0.0.0.0:8080", client=MnistClient())
+
+    # print("training done")
```
[the remark this answers is not in the document]
Comment (fd0r) [reply use_case_examples/federated_learning/client.py:50]: Yes indeed

File: use_case_examples/federated_learning/load_to_cml.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import pickle
+
+import numpy
+import utils
+
+from concrete.ml.sklearn import LogisticRegression
+
+if __name__ == "__main__":
+    # Load scikit-learn model
+    with open("model.pkl", "rb") as file:
```
[the remark this answers is not in the document]
Comment (fd0r) [reply use_case_examples/federated_learning/load_to_cml.py:10]: Good idea!

File: use_case_examples/federated_learning/load_to_cml.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        sklearn_model = pickle.load(file)
+
+    # Compile model without data since we aren't suppose to have any
```
[the remark this answers is not in the document]
Comment (fd0r) [reply use_case_examples/federated_learning/load_to_cml.py:13]: Alright will do

File: use_case_examples/federated_learning/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Federated Learning
+
+This is an example on how to use a federated learning trained model with Concrete ML.
```
Comment (jfrery) [inline use_case_examples/federated_learning/README.md:3]: I find `a federated learning trained model` a bit hard to read.

Maybe,

```suggestion
This use case example combines both federated learning for training and Fully Homomorphic Encryption (FHE) for inference on encrypted data. This approach allows both training and inference in the privacy preserving settings.
```

File: use_case_examples/federated_learning/README.md [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+This example is based on the example from the [flower blog post on LogisticRegression training using federated learning with flower](https://flower.dev/blog/2021-07-21-federated-scikit-learn-using-flower/).
```
Comment (jfrery) [inline use_case_examples/federated_learning/README.md:4]:
```suggestion
This example is inspired from the [flower blog post on LogisticRegression training using federated learning with flower](https://flower.dev/blog/2021-07-21-federated-scikit-learn-using-flower/).
```

File: use_case_examples/federated_learning/README.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+[Federated learning files source.](https://github.com/adap/flower/tree/9ee473152f2fde2a5b05b99829db27a607dc46ec/examples/sklearn-logreg-mnist)
+
+## Setup
+
+To run this example you will need to have `concrete-ml` installed and also install the extra requirements from `requirements.txt`, using `python -m pip install -r requirements.txt`.
```
Comment (jfrery) [inline use_case_examples/federated_learning/README.md:9]:
```suggestion
To run this example you will need to have `concrete-ml` installed as well as the extra requirements from `requirements.txt`, using `python -m pip install -r requirements.txt`.
```

File: use_case_examples/federated_learning/README.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+If you are using macOS with arm then you might need to launch the `./fix_grpc.sh` script for fix the `grpcio` package installation.
+
+## Train the model
+
+To train the model just use the `run.sh` script from the blogpost.
```
Comment (jfrery) [inline use_case_examples/federated_learning/README.md:14]: From the blogpost? It's in this directory, no?

**Review by jfrery — COMMENTED** [review]
Thanks. A few comments.

File: use_case_examples/federated_learning/README.md (hunk 7 above)
Comment (fd0r) [reply use_case_examples/federated_learning/README.md:14]: Yes in this directory but the script itself was copied from the files of the blogpost.
I see how that can be confusing, I'll change that.

**Review by jfrery — APPROVED** [review]

File: use_case_examples/federated_learning/federated_utils.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+        ]
+    else:
+        params = [
+            model.coef_,
+        ]
+    return params
+
+
+def set_model_params(model: LogisticRegression, params: LogRegParams) -> LogisticRegression:
+    """Sets the parameters of a sklean LogisticRegression model."""
```
Comment (RomanBredehoft) [inline use_case_examples/federated_learning/federated_utils.py:231]: `sklean` -> `sklearn`

Comment (fd0r) [reply use_case_examples/federated_learning/federated_utils.py:231]: Damn sorry I merged before waiting.
I'll do a follow up PR to fix the typo.

File: use_case_examples/federated_learning/load_to_cml.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    path_to_model = Path("./model.pkl")
+    if not path_to_model.exists() or not path_to_model.is_file():
+        raise ValueError(
+            "Couldn't find file 'model.pkl'.\n"
+            "Please run 'run.sh' to train the model with federated learning."
+        )
+    with path_to_model.open("rb") as file:
+        sklearn_model = pickle.load(file)
+
+    # Compile model without data since the server doesn't have acces to it
```
Comment (RomanBredehoft) [inline use_case_examples/federated_learning/load_to_cml.py:20]: `acces` -> `access`

**Comment by fd0r** [comment]
https://github.com/zama-ai/concrete-ml/pull/302 @RomanBredehoft

File: use_case_examples/federated_learning/load_to_cml.py (hunk 9 above)
Comment (bcm-at-zama) [reply use_case_examples/federated_learning/load_to_cml.py:20]: hasn't been fixed @fd0r

**Pull request merged**